plugins/pixart-rf/fu-pxi-wireless-device.c

```diff
@@ -0,0 +1,606 @@
+}
+
+static gboolean
+fu_pxi_wireless_device_get_cmd_response (FuPxiWirelessDevice *device,
+					 guint8 *buf, guint bufsz,
+					 GError **error)
+{
+	FuPxiReceiverDevice *parent = fu_pxi_wireless_device_get_parent (FU_DEVICE (device) ,error);
+	guint16 retry = 0;
+	guint8 status = 0x0;
+
+	while (1) {
+		guint8 sn = 0x0;
+		memset (buf, 0, bufsz);
+		buf[0] = PXI_HID_WIRELESS_DEV_OTA_REPORT_ID;
+
+		g_usleep (5 * 1000);
+
+		if (!fu_pxi_wireless_device_get_feature (FU_DEVICE (parent), buf, bufsz, error))
+			return FALSE;
+
+		if (!fu_common_read_uint8_safe (buf, bufsz, 0x4, &sn, error))
+			return FALSE;
+
+		if (device->sn != sn)
+			retry++;
+		else
+			break;
+
+		if (retry == FU_PXI_WIRELESS_DEVICE_RETRY_MAXIMUM) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_READ,
+				     "reach retry maximum hid sn fail, got 0x%04x, expected 0x%04x",
+				     sn, device->sn);
+			return FALSE;
+		}
+
+		/*if wireless device not reply to receiver, keep to wait */
+		if (!fu_common_read_uint8_safe (buf, bufsz, 0x5, &status, error))
+			return FALSE;
+		if (status == OTA_RSP_NOT_READY) {
+			retry = 0x0;
+			g_debug ("OTA_RSP_NOT_READY");
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_wireless_device_check_crc (FuDevice *device, guint16 checksum, GError **error)
+{
+	FuPxiReceiverDevice *parent = fu_pxi_wireless_device_get_parent (device ,error);
+	FuPxiWirelessDevice *self = FU_PXI_WIRELESS_DEVICE (device);
+	guint8 buf[FU_PXI_RECEIVER_DEVICE_OTA_BUF_SZ] = { 0x0 };
+	guint16 checksum_device = 0x0;
+	guint8 status = 0x0;
+	g_autoptr(GByteArray) receiver_cmd = g_byte_array_new ();
+	g_autoptr(GByteArray) ota_cmd = g_byte_array_new ();
+
+	/* ota check crc command */
+	fu_byte_array_append_uint8 (ota_cmd, 0x3);					/* ota command length */
+	fu_byte_array_append_uint8 (ota_cmd, FU_PXI_DEVICE_CMD_FW_OTA_CHECK_CRC);	/* ota command */
+	fu_byte_array_append_uint16 (ota_cmd, checksum, G_LITTLE_ENDIAN);		/* checkesum */
+
+	/* increase the serial number */
+	self->sn++;
+
+	if (!fu_pxi_composite_receiver_cmd (FU_PXI_DEVICE_CMD_FW_OTA_CHECK_CRC,
+					    self->sn,
+					    self->model.target,
+					    receiver_cmd,
+					    ota_cmd,
+					    error))
+		return FALSE;
+	if (!fu_pxi_wireless_device_set_feature (FU_DEVICE (parent),
+						 receiver_cmd->data,
+						 receiver_cmd->len,
+						 error))
+		return FALSE;
+	if (!fu_pxi_wireless_device_get_cmd_response (self, buf, sizeof(buf), error))
+		return FALSE;
+	fu_common_dump_raw (G_LOG_DOMAIN, "crc buf", buf, sizeof(buf));
+	if (!fu_common_read_uint8_safe (buf, sizeof(buf), 0x5, &status, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (buf, sizeof(buf), 0x6,
+					 &checksum_device, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (status == OTA_RSP_CODE_ERROR) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "checksum fail, got 0x%04x, expected 0x%04x",
+			     checksum_device,
+			     checksum);
+		return FALSE;
+	}
+
```

Comment:
shouldn't you also explicitly check for `OTA_RSP_OK`?

Reply:
Add check `OTA_RSP_OK`